### PR TITLE
[DOCS] Added documentation about --asm flag

### PR DESCRIPTION
**Pull request by hrkrshnn**
Context: we get several questions on why two different expression produce different bytecodes. The users usually look at the binary and are clueless. Added some documentation about `--asm`.

### CODE REVIEW

File: docs/index.rst [hunk 1]
```diff
@@ -151,6 +151,7 @@ Contents
    smtchecker.rst
    resources.rst
    using-the-compiler.rst
+   analysing-compilation-output.rst
```
Comment (chriseth) [inline docs/index.rst]: Do we really want this as a new top-level element? Shouldn't it be part of "using the compiler"?

File: docs/analysing-compilation-output.rst [hunk 2]
```diff
@@ -0,0 +1,190 @@
+.. index:: analyse, asm
+
+#############################
+Analysing the Compiler Output
+#############################
+
+It is often useful to look at the assembly code generated by the compiler. However, the generated
```
Comment (chriseth) [inline docs/analysing-compilation-output.rst]: Why "however"? This sounds like `--bin` would create "assembly code".

File: docs/analysing-compilation-output.rst [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+binary, i.e., the output of ``solc --bin contract.sol``, is difficult to read. It is recommended to use
+the flag ``--asm`` to analyse the assembly output.
```
Comment (chriseth) [inline docs/analysing-compilation-output.rst]: Maybe add: Even for large contracts, looking a a graphical diff of the assembly before and after a change is often very enlightening.

File: docs/analysing-compilation-output.rst [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+
+Consider the following contract (named, say ``contract.sol``):
+
+.. code-block:: Solidity
+
+    // SPDX-License-Identifier: GPL-3.0
+    pragma solidity >= 0.8.0 < 0.9.0;
```
Comment (chriseth) [inline docs/analysing-compilation-output.rst]:
```suggestion
    pragma solidity ^0.8.0;
```

Comment (chriseth) [reply docs/analysing-compilation-output.rst]: Actually you can (and should) go back with the min version requirement a lot.

File: docs/analysing-compilation-output.rst [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+            return 1;
+        }
+    }
+
+The following would be the output of ``solc --asm contract.sol``
+
+
+.. code-block:: none
+
+    ======= /tmp/test.sol:C =======
```
Comment (chriseth) [inline docs/analysing-compilation-output.rst]: Can you compile it so that it just says `test.sol`?

File: docs/analysing-compilation-output.rst [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+Alternatively, the above output can also be obtained from `Remix <https://remix.ethereum.org/>`_,
+under the option "Compilation Details" after compiling a contract.
+
+Notice that the ``asm`` output starts with the creation / constructor code. The deploy code is
+provided as part of the sub object (in the above example, it is part of the sub-object ``sub_0``).
+The ``auxdata`` field corresponds to the contract :ref:`metadata
+<encoding-of-the-metadata-hash-in-the-bytecode>`.
+
+Similarly, the optimized assembly can be obtained with the command: ``solc --optimize --asm
+contract.sol``. Often times, it is interesting to see if two different code in Solidity results in
```
Comment (chriseth) [inline docs/analysing-compilation-output.rst]:
```suggestion
contract.sol``. Often times, it is interesting to see if two different sources in Solidity result in
```

File: docs/analysing-compilation-output.rst [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+
+
+Alternatively, the above output can also be obtained from `Remix <https://remix.ethereum.org/>`_,
+under the option "Compilation Details" after compiling a contract.
+
+Notice that the ``asm`` output starts with the creation / constructor code. The deploy code is
+provided as part of the sub object (in the above example, it is part of the sub-object ``sub_0``).
+The ``auxdata`` field corresponds to the contract :ref:`metadata
+<encoding-of-the-metadata-hash-in-the-bytecode>`.
+
```
Comment (chriseth) [inline docs/analysing-compilation-output.rst:182]: Can you explain that the comments tell you which source part the code was generated for and that `#utility.yul` is an internally-generated file of utility functions that can be obtained using `--combined-json generated-sources,generated-sources-runtime`?

**Review by chriseth — CHANGES_REQUESTED** [review]

File: docs/index.rst (hunk 1 above)
Comment (hrkrshnn) [reply docs/index.rst]: Not sure. `using-the-compiler` is a rather long already.

File: docs/analysing-compilation-output.rst (hunk 4 above)
Comment (hrkrshnn) [reply docs/analysing-compilation-output.rst]: Changed to 0.5.0

File: docs/analysing-compilation-output.rst (hunk 7 above)
Comment (hrkrshnn) [reply docs/analysing-compilation-output.rst:182]: Added this.

File: docs/analysing-compilation-output.rst [hunk 8]
```diff
@@ -0,0 +1,192 @@
+.. index:: analyse, asm
+
+#############################
+Analysing the Compiler Output
```
Comment (bshastry) [inline docs/analysing-compilation-output.rst:4]: I'm always confused if we should use the American or British spelling `analyzing` vs `analysing` :-)

File: docs/analysing-compilation-output.rst [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+#############################
+
+It is often useful to look at the assembly code generated by the compiler. The generated binary,
+i.e., the output of ``solc --bin contract.sol``, is generally difficult to read. It is recommended
+to use the flag ``--asm`` to analyse the assembly output. Even for large contracts, looking at a
+graphical diff of the assembly before and after a change is often very enlightening.
```
Comment (bshastry) [inline docs/analysing-compilation-output.rst]: `visual` instead of `graphical` diff?

File: docs/analysing-compilation-output.rst [hunk 10]
```diff
@@ -0,0 +1,192 @@
...
+        auxdata: 0xa2646970667358221220a5874f19737ddd4c5d77ace1619e5160c67b3d4bedac75fce908fed32d98899864736f6c637827302e382e342d646576656c6f702e323032312e332e33302b636f6d6d69742e65613065363933380058
+    }
+
+Alternatively, the above output can also be obtained from `Remix <https://remix.ethereum.org/>`_,
+under the option "Compilation Details" after compiling a contract.
+
+Notice that the ``asm`` output starts with the creation / constructor code. The deploy code is
+provided as part of the sub object (in the above example, it is part of the sub-object ``sub_0``).
+The ``auxdata`` field corresponds to the contract :ref:`metadata
+<encoding-of-the-metadata-hash-in-the-bytecode>`. The comments in the assembly output points to the
```
Comment (bshastry) [inline docs/analysing-compilation-output.rst]:
```suggestion
<encoding-of-the-metadata-hash-in-the-bytecode>`. The comments in the assembly output point to the
```

**Review by bshastry — COMMENTED** [review]
Minor comments. I think the main thing to be resolved in this PR if the documentation should be top level or not

File: docs/analysing-compilation-output.rst (hunk 8 above)
Comment (hrkrshnn) [reply docs/analysing-compilation-output.rst:4]: I think docs use British spelling. So `analyse` should work, right?

File: docs/analysing-compilation-output.rst (hunk 10 above)
Comment (hrkrshnn) [reply docs/analysing-compilation-output.rst]: fixed both the comments.

File: docs/index.rst (hunk 1 above)
Comment (hrkrshnn) [reply docs/index.rst]: Also, we could extend this section in the future to include the `ir` output. So, I would lean towards a top level element.

Comment (chriseth) [reply docs/index.rst]: We could have a new big section next to "language description" and "internals" called "compiler"

File: docs/analysing-compilation-output.rst [hunk 11]
```diff
@@ -0,0 +1,192 @@
...
+<encoding-of-the-metadata-hash-in-the-bytecode>`. The comments in the assembly output point to the
+source location. Note that ``#utility.yul`` is an internally generated file of utility functions
+that can be obtained using the flags ``--combined-json
+generated-sources,generated-sources-runtime``.
+
+Similarly, the optimized assembly can be obtained with the command: ``solc --optimize --asm
+contract.sol``. Often times, it is interesting to see if two different sources in Solidity result in
+the same optimized code. For example, to see if the expressions ``(a * b) / c``, ``a * b / c``
+generates the same bytecode. This can be easily done by taking a ``diff`` of the corresponding
+assembly output.
```
Comment (chriseth) [inline docs/analysing-compilation-output.rst]: Potentially after stripping the comments that reference the source location.

File: docs/analysing-compilation-output.rst [hunk 12]
```diff
@@ -0,0 +1,192 @@
...
+source location. Note that ``#utility.yul`` is an internally generated file of utility functions
+that can be obtained using the flags ``--combined-json
+generated-sources,generated-sources-runtime``.
+
+Similarly, the optimized assembly can be obtained with the command: ``solc --optimize --asm
+contract.sol``. Often times, it is interesting to see if two different sources in Solidity result in
+the same optimized code. For example, to see if the expressions ``(a * b) / c``, ``a * b / c``
+generates the same bytecode. This can be easily done by taking a ``diff`` of the corresponding
+assembly output, after potentially stripping comments that reference the source locations.
+
```
Comment (hrkrshnn) [inline docs/analysing-compilation-output.rst:188]: Added this line.

File: docs/index.rst [hunk 13]
```diff
@@ -127,6 +127,13 @@ Contents
    cheatsheet.rst
    grammar.rst
 
+.. toctree::
+   :maxdepth: 2
+   :caption: Compiler
+
+   using-the-compiler.rst
+   analysing-compilation-output.rst
```
Comment (hrkrshnn) [inline docs/index.rst:135]: Added this new big section.

**Review by bshastry — APPROVED** [review]
LGTM. Is a preview available, so one can see how the diff is rendered?

**Comment by hrkrshnn** [comment]
@bshastry https://693059-40892817-gh.circle-artifacts.com/0/docs-html/analysing-compilation-output.html.

It's from the `b_docs` CI job.

File: docs/index.rst (hunk 13 above)
Comment (chriseth) [reply docs/index.rst:135]: I think some other pages also fall into this section, but let's leave it like that for now.

**Review by chriseth — APPROVED** [review]

**Pull request merged**